### ISSUE TITLE
File Locking in Windows(Rolling logs, etc.)

### ISSUE THREAD

**Issue by futurechan** [elasticsearch/logstash#1557]
https://logstash.jira.com/browse/LOGSTASH-986

Has a fix for this been merged yet?

**Comment by deancsmith** [comment]
+1 this is causing huge problems for me on Win 2008r2 servers.

The workaround is not really practical as we are monitoring a dozen or so files in several dozen directories (including IIS)where we won't always know the file name.

**Comment by ThomasGoeytil** [comment]
+1 on this question!

**Comment by futurechan** [comment]
LogCourier is a practical work around to get log files into logstash. We've switched all of our apps to log to logstash over tcp.

**Comment by jordansissel** [comment]
I'm pretty sure this is fixed in the latest filewatch library, but I will be confirming before Logstash 1.5.0 is released.

**Comment by futurechan** [comment]
I am definitely pumped for 1.5

**Comment by ppiersonbt** [comment]
Not sure if I am tracking this right or not but would this explain why I can't pull IIS logs via logstash?

**Comment by deancsmith** [comment]
@ppiersonbt You can process IIS logs in Logstash, you just have to configure it a little bit, the above issue is just to do with the file locking of these logs. (causes problems when Logstash is restarted or when rolling over logs)

Here is my logstash configuration for IIS. It runs on multiple servers but I use the same config file on all servers.

```
input{
    file {
    path => [ "D:/logs/iis/**/*.log" ]
    type => "iis"
  }
}
filter {
  if [type] == "iis" {
    # Ignore IIS log comments
      if [message] =~ "^#" {
      drop {}
    }

    # Set the event timestamp from the log
    date {
     match => [ "log_timestamp", "YYYY-MM-dd HH:mm:ss" ]
     #timezone => "Etc/UCT"
     }

    grok {
      match => ["message", "%{TIMESTAMP_ISO8601:log_timestamp} %{WORD:iisSite} %{WORD:iisComputerName} %{IPORHOST:site} %{WORD:method} %{URIPATH:page} %{NOTSPACE:querystring} %{NUMBER:port} %{NOTSPACE:username} %{IPORHOST:clienthost} %{NOTSPACE:useragent} %{NOTSPACE:referer} %{NUMBER:response} %{NUMBER:subresponse} %{NUMBER:scstatus} %{NUMBER:bytes} %{NUMBER:timetaken}"]
    }

    # Convert plus signs in the user agent to a space before processing them into their separate fields
    if [useragent] =~ "[A-Za-z]+" {
      mutate {
        gsub => [
          # Replace plus sign with space
          "useragent", "\+", " "
        ]
      }
      useragent {
        source=> "useragent"
        prefix=> "browser"
      }
    }		
    
    # Convert most columns to int or float, depending on what they contain for better Kibana searching
    mutate {
      remove_field => [ "log_timestamp"]
      convert => [ "bytes", "integer" ]
      convert => [ "response", "integer" ]
      convert => [ "scstatus", "integer" ]
      convert => [ "timetaken", "integer" ]
      # Remove fields we don't need to reduce overall data
      remove_field => [ "path", "message" ]
    }
    # Replace IIS Site IDs with application names.
    # We put the config for all servers here so it can be the same in all Logstash configs
    if [host] == "SERVER1" {
      if [iisSite] == "W3SVC2"{
        mutate { replace => [ "iisSite", "app1" ] }
      }
      if [iisSite] == "W3SVC3"{
        mutate { replace => [ "iisSite", "app2" ] }
      }
    }
    if [host] == "SERVER2" {
      if [iisSite] == "W3SVC2"{
        mutate { replace => [ "iisSite", "app3" ] }
      }
      if [iisSite] == "W3SVC3"{
        mutate { replace => [ "iisSite", "app4" ] }
      }
  }
}
output {
  elasticsearch_http {
    host => "server.foo.com"
  }
}
```

Please note that we use a custom logging format for our IIS sites (configure through the Logging section in the IIS site, or we configure all our servers using Chef) to change the log path and include extra fields (e.g. size, user agent and a few other things). Here are the relevant logging bits:

```
# Set the log path
iis_config "/section:sites /[name='#{ENV['SETUP_WEB_SITE_NAME']}'].logFile.directory:\"#{ENV['SD_LOG_PATH']}\iis\"" do
  action :config
end

# Then set the log fields that we want
iis_config "/section:sites /[name='#{ENV['SETUP_WEB_SITE_NAME']}'].logFile.logExtFileFlags:\"Date, Time, ClientIP, UserName, SiteName, ComputerName, ServerIP, Method, UriStem, UriQuery, HttpStatus, Win32Status, BytesSent, TimeTaken, ServerPort, UserAgent, Referer, HttpSubStatus\"" do
  action :config
end
```

**Comment by ppiersonbt** [comment]
yeah sorry @deancsmith this put me down the same path i was down before.  I did find out what the issue was but I doubt anyone will be able to replicate it.  On my windows 2008 r2 vm, if I do not have     start_position => "beginning"  before the file entry it doesn't work.  If i place start_position after the file entry the logs won't load.  Thanks again for the help.  If I am able to find a way to replicate this issue outside of the one vm so far I will report it.

**Comment by tkorkalainen** [comment]
+1

**Comment by suyograo** [comment]
Fixed in https://github.com/jordansissel/ruby-filewatch/pull/16 @jsvd released 0.6.1 version of fielwatch gem which will ship with LS 1.5 RC1
